src/execution/sql/sorter.cpp

```diff
@@ -0,0 +1,343 @@
+      child++;
+    }
+
+    if (cmp_fn_(top, tuples_[child]) >= 0) {
+      break;
+    }
+
+    std::swap(tuples_[idx], tuples_[child]);
+    idx = child;
+  }
+
+  tuples_[idx] = top;
+}
+
+void Sorter::Sort() {
+  // Exit if the input tuples have already been sorted
+  if (is_sorted()) {
+    return;
+  }
+
+  // Exit if there are no input tuples
+  if (tuples_.empty()) {
+    return;
+  }
+
+  // Time it
+  util::Timer<std::milli> timer;
+  timer.Start();
+
+  // Sort the sucker
+  const auto compare = [this](const byte *left, const byte *right) { return cmp_fn_(left, right) < 0; };
+  ips4o::sort(tuples_.begin(), tuples_.end(), compare);
+
+  timer.Stop();
+
+  UNUSED double tps = (static_cast<double>(tuples_.size()) / timer.elapsed()) / 1000.0;
+  EXECUTION_LOG_DEBUG("Sorted {} tuples in {} ms ({:.2f} tps)", tuples_.size(), timer.elapsed(), tps);
+
+  // Mark complete
+  sorted_ = true;
+}
+
+namespace {
+
+// Structure we use to track a package of merging work.
+template <typename IterType>
+struct MergeWork {
+  using Range = std::pair<IterType, IterType>;
+
+  std::vector<Range> input_ranges;
+  IterType destination;
+
+  MergeWork(std::vector<Range> &&inputs, IterType dest) : input_ranges(std::move(inputs)), destination(dest) {}
+};
+
+}  // namespace
+
+void Sorter::SortParallel(const ThreadStateContainer *thread_state_container, const u32 sorter_offset) {
+  const auto comp = [this](const byte *left, const byte *right) { return cmp_fn_(left, right) < 0; };
+
+  // -------------------------------------------------------
+  // First, collect all non-empty thread-local sorters
+  // -------------------------------------------------------
+
+  std::vector<Sorter *> tl_sorters;
+  thread_state_container->CollectThreadLocalStateElementsAs(&tl_sorters, sorter_offset);
+  llvm::erase_if(tl_sorters, [](Sorter *const sorter) { return sorter->NumTuples() == 0; });
+
+  // If there's nothing to sort, quit
+  if (tl_sorters.empty()) {
+    sorted_ = true;
+    return;
+  }
+
+  // -------------------------------------------------------
+  // 1. Make room in this sorter for all result tuples
+  // -------------------------------------------------------
+
+  util::StageTimer<std::milli> timer;
+  timer.EnterStage("Resize Main Sorter");
+
+  const u64 num_tuples =
+      std::accumulate(tl_sorters.begin(), tl_sorters.end(), u64(0),
+                      [](const u64 partial, const Sorter *const sorter) { return partial + sorter->NumTuples(); });
+  tuples_.resize(num_tuples);
+
+  timer.ExitStage();
+
+  // -------------------------------------------------------
+  // 2. Sort each thread-local sorter in parallel
+  // -------------------------------------------------------
+
+  timer.EnterStage("Parallel Sort Thread-Local Instances");
+
+  tbb::task_scheduler_init sched;
+  tbb::parallel_for_each(tl_sorters.begin(), tl_sorters.end(), [](Sorter *const sorter) { sorter->Sort(); });
```

Comment:
@mbutrovich and @pmenon, we've agreed could be a problem since we haven't unified thread management.  What's the way forward on this one?  Accept as is and open an issue or solve the thread pool problem as part of this PR?